In [2]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
import seaborn as sns
import matplotlib.pyplot as plt
from mlxtend.feature_selection import SequentialFeatureSelector as SFS

from sklearn.datasets import make_classification
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import RFECV
from sklearn.model_selection import StratifiedKFold

from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.feature_selection import RFE

from sklearn.datasets import load_iris
from sklearn.model_selection import StratifiedKFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import RFECV

from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.svm import SVC

In [3]:
data = pd.read_csv(r'C:\Users\장건호\Desktop\BDA\taiwanese+bankruptcy+prediction\data.csv')

In [4]:
#sfs_forward 선택된 특성
selected_1 = [' ROA(C) before interest and depreciation before interest', ' ROA(B) before interest and depreciation after tax', ' Operating Gross Margin', ' Realized Sales Gross Margin', ' Operating Profit Rate', ' Pre-tax net Interest Rate', ' After-tax net Interest Rate', ' Non-industry income and expenditure/revenue', ' Continuous interest rate (after tax)', ' Cash flow rate', ' Net Value Per Share (B)', ' Net Value Per Share (A)', ' Net Value Per Share (C)', ' Persistent EPS in the Last Four Seasons', ' Cash Flow Per Share', ' Revenue Per Share (Yuan ¥)', ' Operating Profit Per Share (Yuan ¥)', ' Per Share Net profit before tax (Yuan ¥)', ' Realized Sales Gross Profit Growth Rate', ' Operating Profit Growth Rate', ' Borrowing dependency', ' Long-term Liability to Current Assets', ' Equity to Long-term Liability']
X1 = data[selected_1]
#sfs_backward 선택된 특성
selected_2 = [' ROA(C) before interest and depreciation before interest', ' ROA(A) before interest and % after tax']
X2 = data[selected_2]
#sfs_stepwise 선택된 특성
selected_3 = [' ROA(C) before interest and depreciation before interest', ' ROA(A) before interest and % after tax', ' ROA(B) before interest and depreciation after tax', ' Operating Gross Margin', ' Realized Sales Gross Margin', ' Operating Profit Rate', ' Pre-tax net Interest Rate', ' After-tax net Interest Rate', ' Non-industry income and expenditure/revenue', ' Continuous interest rate (after tax)', ' Cash flow rate', ' Tax rate (A)', ' Net Value Per Share (B)', ' Net Value Per Share (A)', ' Net Value Per Share (C)', ' Persistent EPS in the Last Four Seasons', ' Cash Flow Per Share', ' Revenue Per Share (Yuan ¥)', ' Operating Profit Per Share (Yuan ¥)', ' Per Share Net profit before tax (Yuan ¥)', ' Realized Sales Gross Profit Growth Rate', ' Operating Profit Growth Rate', ' After-tax Net Profit Growth Rate', ' Regular Net Profit Growth Rate', ' Continuous Net Profit Growth Rate', ' Total Asset Return Growth Rate Ratio', ' Cash Reinvestment %', ' Current Ratio', ' Interest Expense Ratio', ' Debt ratio %', ' Net worth/Assets', ' Long-term fund suitability ratio (A)', ' Borrowing dependency', ' Contingent liabilities/Net worth', ' Operating profit/Paid-in capital', ' Net profit before tax/Paid-in capital', ' Inventory and accounts receivable/Net value', ' Total Asset Turnover', ' Net Worth Turnover Rate (times)', ' Operating profit per person', ' Working Capital to Total Assets', ' Quick Assets/Total Assets', ' Current Assets/Total Assets', ' Cash/Total Assets', ' Quick Assets/Current Liability', ' Current Liability to Assets', ' Operating Funds to Liability', ' Inventory/Working Capital', ' Current Liabilities/Liability', ' Working Capital/Equity', ' Current Liabilities/Equity', ' Retained Earnings to Total Assets', ' Total income/Total expense', ' Total expense/Assets', ' Working capitcal Turnover Rate', ' Cash Flow to Sales', ' Current Liability to Liability', ' Current Liability to Equity', ' Equity to Long-term Liability', ' Cash Flow to Total Assets', ' Cash Flow to Liability', ' CFO to Assets', ' Cash Flow to Equity', ' Current Liability to Current Assets', ' Net Income to Total Assets', ' No-credit Interval', ' Gross Profit to Sales', " Net Income to Stockholder's Equity", ' Degree of Financial Leverage (DFL)', ' Interest Coverage Ratio (Interest expense to EBIT)', ' Net Income Flag', ' Equity to Liability']
X3 = data[selected_3]
#RFE 선택된 특성
selected_4 = [' Revenue Per Share (Yuan ¥)', ' Accounts Receivable Turnover',
       ' Average Collection Days', ' Allocation rate per person',
       ' Fixed Assets to Assets']
X4 = data[selected_4]
#RFECV 선택된 특성
selected_5 = [' ROA(C) before interest and depreciation before interest',
       ' Non-industry income and expenditure/revenue',
       ' Continuous interest rate (after tax)',
       ' Interest-bearing debt interest rate', ' Net Value Per Share (C)',
       ' Persistent EPS in the Last Four Seasons',
       ' Per Share Net profit before tax (Yuan ¥)', ' Net Value Growth Rate',
       ' Quick Ratio', ' Interest Expense Ratio',
       ' Total debt/Total net worth', ' Net worth/Assets',
       ' Borrowing dependency', ' Net profit before tax/Paid-in capital',
       ' Accounts Receivable Turnover', ' Cash/Total Assets',
       ' Inventory/Working Capital', ' Working Capital/Equity',
       ' Net Income to Total Assets', ' Net Income to Stockholder\'s Equity',
       ' Degree of Financial Leverage (DFL)',
       ' Interest Coverage Ratio (Interest expense to EBIT)']
X5 = data[selected_5]

y = data['Bankrupt?']

In [6]:
models = {
    'LogisticRegression': LogisticRegression(max_iter=200),
    #'RandomForestClassifier': RandomForestClassifier(),
    #'SVC': SVC(kernel='linear')
}

# 평가 지표를 계산하는 함수
def evaluate_model(X, y, model):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, zero_division=0)
    recall = recall_score(y_test, y_pred, zero_division=0)
    f1 = f1_score(y_test, y_pred, zero_division=0)
    
    return accuracy, precision, recall, f1

# 각 피처 선택 방법과 모델에 대한 평가 지표 계산
results = []

for model_name, model in models.items():
    for i, (X, selected_features) in enumerate(zip([X1, X2, X3, X4, X5], 
                                                   [selected_1, selected_2, selected_3, selected_4, selected_5]), 1):
        accuracy, precision, recall, f1 = evaluate_model(X, y, model)
        results.append({
            'Model': model_name,
            'Feature Selection Method': f'selected_{i}',
            'Accuracy': accuracy,
            'Precision': precision,
            'Recall': recall,
            'F1': f1,
            'Selected Features': selected_features
        })

# 결과 출력
results_df = pd.DataFrame(results)
print(results_df)



                Model Feature Selection Method  Accuracy  Precision    Recall  \
0  LogisticRegression               selected_1  0.961877   0.000000  0.000000   
1  LogisticRegression               selected_2  0.961877   0.500000  0.025641   
2  LogisticRegression               selected_3  0.964321   0.777778  0.089744   
3  LogisticRegression               selected_4  0.958456   0.000000  0.000000   
4  LogisticRegression               selected_5  0.961877   0.000000  0.000000   

        F1                                  Selected Features  
0  0.00000  [ ROA(C) before interest and depreciation befo...  
1  0.04878  [ ROA(C) before interest and depreciation befo...  
2  0.16092  [ ROA(C) before interest and depreciation befo...  
3  0.00000  [ Revenue Per Share (Yuan ¥),  Accounts Receiv...  
4  0.00000  [ ROA(C) before interest and depreciation befo...  


c:\Python311\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=2):
ABNORMAL_TERMINATION_IN_LNSRCH.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [7]:
# 모델 정의
models = {
    'LogisticRegression': LogisticRegression(max_iter=200),
    # 'RandomForestClassifier': RandomForestClassifier(),
    # 'SVC': SVC(kernel='linear')
}

# 평가 지표를 계산하는 함수
def evaluate_model(X, y, model):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=111)
    model.fit(X_train, y_train)
    y_pred_train = model.predict(X_train)
    y_pred_test = model.predict(X_test)
    
    train_metrics = {
        'Accuracy': accuracy_score(y_train, y_pred_train),
        'Precision': precision_score(y_train, y_pred_train, zero_division=0),
        'Recall': recall_score(y_train, y_pred_train, zero_division=0),
        'F1': f1_score(y_train, y_pred_train, zero_division=0)
    }
    
    test_metrics = {
        'Accuracy': accuracy_score(y_test, y_pred_test),
        'Precision': precision_score(y_test, y_pred_test, zero_division=0),
        'Recall': recall_score(y_test, y_pred_test, zero_division=0),
        'F1': f1_score(y_test, y_pred_test, zero_division=0)
    }
    
    return train_metrics, test_metrics

# 각 피처 선택 방법과 모델에 대한 평가 지표 계산
results = []

for model_name, model in models.items():
    for i, (X, selected_features) in enumerate(zip([X1, X2, X3, X4, X5], 
                                                   [selected_1, selected_2, selected_3, selected_4, selected_5]), 1):
        train_metrics, test_metrics = evaluate_model(X, y, model)
        
        results.append({
            'Model': model_name,
            'Dataset': f'selected_{i}',
            'Train Accuracy': train_metrics['Accuracy'],
            'Train Precision': train_metrics['Precision'],
            'Train Recall': train_metrics['Recall'],
            'Train F1': train_metrics['F1'],
            'Test Accuracy': test_metrics['Accuracy'],
            'Test Precision': test_metrics['Precision'],
            'Test Recall': test_metrics['Recall'],
            'Test F1': test_metrics['F1'],
            'Selected Features': ', '.join(selected_features)
        })

# 결과 출력
results_df = pd.DataFrame(results)
print(results_df)


                Model     Dataset  Train Accuracy  Train Precision  \
0  LogisticRegression  selected_1        0.967919         0.000000   
1  LogisticRegression  selected_2        0.967736         0.461538   
2  LogisticRegression  selected_3        0.970119         0.730769   
3  LogisticRegression  selected_4        0.222548         0.031351   
4  LogisticRegression  selected_5        0.967186         0.166667   

   Train Recall  Train F1  Test Accuracy  Test Precision  Test Recall  \
0      0.000000  0.000000       0.967009        0.000000     0.000000   
1      0.034286  0.063830       0.967009        0.500000     0.044444   
2      0.108571  0.189055       0.968475        0.625000     0.111111   
3      0.777143  0.060270       0.207478        0.035009     0.866667   
4      0.005714  0.011050       0.966276        0.000000     0.000000   

    Test F1                                  Selected Features  
0  0.000000   ROA(C) before interest and depreciation befor...  
1  0.08163

In [13]:
from collections import Counter
total_selected = selected_1 + selected_2 + selected_3 + selected_4 + selected_5
total_number = Counter(total_selected)
selected = [elem for elem, count in total_number.items() if count > 2]

selected

[' ROA(C) before interest and depreciation before interest',
 ' Non-industry income and expenditure/revenue',
 ' Continuous interest rate (after tax)',
 ' Net Value Per Share (C)',
 ' Persistent EPS in the Last Four Seasons',
 ' Revenue Per Share (Yuan ¥)',
 ' Per Share Net profit before tax (Yuan ¥)',
 ' Borrowing dependency']

In [26]:
corr = [ ' Debt ratio %', ' Current Liability to Assets',
       ' Borrowing dependency', ' Current Liability to Current Assets',
       ' Liability to Equity', ' Current Liabilities/Equity',
       ' Current Liability to Equity', ' Liability-Assets Flag',
       ' Total expense/Assets', ' Equity to Long-term Liability']

In [32]:
!pip install imblearn

                                              0.0/258.0 kB ? eta -:--:--
     -----                                 41.0/258.0 kB 960.0 kB/s eta 0:00:01
     ------------------                     122.9/258.0 kB 1.4 MB/s eta 0:00:01
     ------------------------------         204.8/258.0 kB 1.8 MB/s eta 0:00:01
     -------------------------------------- 258.0/258.0 kB 1.8 MB/s eta 0:00:00



[notice] A new release of pip is available: 23.1.2 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [33]:
from imblearn.over_sampling import SMOTE

In [ ]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, recall_score, f1_score
from sklearn.pipeline import make_pipeline

# Support Vector Classifier (SVC) 모델
svc_model = SVC()
svc_model = make_pipeline(StandardScaler(), SVC(class_weight='balanced'))
svc_model.fit(X_train, y_train)
y_pred_svc = svc_model.predict(X_test)
print("SVC Accuracy:", accuracy_score(y_test, y_pred_svc))
print("SVC recall:\n", recall_score(y_test, y_pred_svc))
print("SVC f1:\n", f1_score(y_test,y_pred_svc))

In [50]:
X = data[corr]
y = data['Bankrupt?']

bankrupt_count = data['Bankrupt?'].value_counts()
max_count = bankrupt_count.max()
smote = SMOTE(sampling_strategy={i: max_count for i in bankrupt_count.index})

X, y = smote.fit_resample(X, y)
balanced_data = pd.concat([X, y], axis=1)

# Check the balanced count for each attack category
balanced_Bankrupt_counts = balanced_data['Bankrupt?'].value_counts()

# 모델 정의
models = {
    #'LogisticRegression': LogisticRegression(max_iter=200)
    'RandomForestClassifier': RandomForestClassifier(n_estimators=1000,  # 기본 트리의 수
        max_depth=5,  # 트리의 최대 깊이 제한
        min_samples_split=2,  # 노드를 분할하는 데 필요한 최소 샘플 수
        min_samples_leaf=1,  # 리프 노드에 있어야 하는 최소 샘플 수
        max_features='sqrt',  # 각 분할에 사용할 피처의 최대 수
        random_state=111  # 결과 재현성을 위한 랜덤 시드
    )
    #'SVC': SVC(StandardScaler(), SVC(class_weight='balanced'))
}

# 평가 지표를 계산하는 함수
def evaluate_model(X, y, model):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=111)
    model.fit(X_train, y_train)
    y_pred_train = model.predict(X_train)
    y_pred_test = model.predict(X_test)
    
    train_metrics = {
        'Accuracy': accuracy_score(y_train, y_pred_train),
        'Precision': precision_score(y_train, y_pred_train, zero_division=0),
        'Recall': recall_score(y_train, y_pred_train, zero_division=0),
        'F1': f1_score(y_train, y_pred_train, zero_division=0)
    }
    
    test_metrics = {
        'Accuracy': accuracy_score(y_test, y_pred_test),
        'Precision': precision_score(y_test, y_pred_test, zero_division=0),
        'Recall': recall_score(y_test, y_pred_test, zero_division=0),
        'F1': f1_score(y_test, y_pred_test, zero_division=0)
    }
    
    return train_metrics, test_metrics

# 모델 평가
results = []

for model_name, model in models.items():
    train_metrics, test_metrics = evaluate_model(X, y, model)
    
    results.append({
        'Model': model_name,
        'Train Accuracy': train_metrics['Accuracy'],
        'Train Precision': train_metrics['Precision'],
        'Train Recall': train_metrics['Recall'],
        'Train F1': train_metrics['F1'],
        'Test Accuracy': test_metrics['Accuracy'],
        'Test Precision': test_metrics['Precision'],
        'Test Recall': test_metrics['Recall'],
        'Test F1': test_metrics['F1'],
        'Selected Features': ', '.join(selected_features)
    })

# 결과 출력
results_df = pd.DataFrame(results)
print(results_df)


                    Model  Train Accuracy  Train Precision  Train Recall  \
0  RandomForestClassifier        0.874597          0.86972      0.880296   

   Train F1  Test Accuracy  Test Precision  Test Recall   Test F1  \
0  0.874976       0.866667        0.868816     0.867515  0.868165   

                                   Selected Features  
0  ROA(C) before interest and depreciation before...  


In [47]:
from sklearn.pipeline import make_pipeline
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

In [48]:
y = data['Bankrupt?']
X = data[selected]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=111)

# Pipeline 정의
pipeline = Pipeline([
    ('scaler', StandardScaler()),  # StandardScaler를 사용하여 데이터 표준화
    ('svc', SVC(class_weight='balanced'))  # SVC 모델을 사용
])

# 모델 학습
pipeline.fit(X_train, y_train)

# 예측
y_pred_train = pipeline.predict(X_train)
y_pred_test = pipeline.predict(X_test)

# 평가 지표 계산
train_metrics = {
    'Accuracy': accuracy_score(y_train, y_pred_train),
    'Precision': precision_score(y_train, y_pred_train, zero_division=0),
    'Recall': recall_score(y_train, y_pred_train, zero_division=0),
    'F1': f1_score(y_train, y_pred_train, zero_division=0)
}

test_metrics = {
    'Accuracy': accuracy_score(y_test, y_pred_test),
    'Precision': precision_score(y_test, y_pred_test, zero_division=0),
    'Recall': recall_score(y_test, y_pred_test, zero_division=0),
    'F1': f1_score(y_test, y_pred_test, zero_division=0)
}

print("Train Metrics:", train_metrics)
print("Test Metrics:", test_metrics)

Train Metrics: {'Accuracy': 0.8612282309807516, 'Precision': 0.1759465478841871, 'Recall': 0.9028571428571428, 'F1': 0.29450139794967384}
Test Metrics: {'Accuracy': 0.8680351906158358, 'Precision': 0.18009478672985782, 'Recall': 0.8444444444444444, 'F1': 0.29687500000000006}


다양한 방법으로 피처를 선정하고, 모델 사용도 다양하게 진행해보았다.
하지만 피처선정과, 모델 선택과 관계없이 전체적인 평가지표가 고르게 나오는 경우가 없어, 데이터 전처리가 필요하다고 판단하였고, Smote Sampling을 통해서 뷸균형 데이터를 완화해서 랜덤포레스트의 파라미터를 변경해가며 학습을 진행시켰더니                     

Model                       Train Accuracy   Train Precision  Train Recall 
RandomForestClassifier        0.874597          0.86972         0.880296   

Train F1  Test Accuracy  Test Precision  Test Recall    Test F1  
0.874976    0.866667        0.868816     0.867515       0.868165 

와 같은 결과를 얻을 수 있었습니다.
